### DISCUSSION TITLE
Can't get localization working...

### DISCUSSION THREAD

**Discussion by tdittmar** [dotnetcore/BootstrapBlazor#1192] [Q&A] [answered]
Hi, I'm currently working on a .NET MAUI Blazor app, which I want to be in German only. Currently, all text coming from BootstrapBlazor is in English, as mentioned in the docs.

I'm now trying to switch everything to German, but I can't seem to get it working:

* I've downloaded the de.json file and placed it in a Locales folder
* I've configured the default locale as "de-DE" in both in appsettings.json and in code
* I'm setting the default culture information to "de-DE"

Still, all texts are in English. Does anyone have a working example on how to switch my entire app to German?

**Comment by ArgoZhang** [comment]
@tdittmar Hi, Could you provider a demo for this?

**Comment by tdittmar** [comment]
I tried so many things, I don't have a "small example", but I can show what I did.

The file `de.json` is in the `Locales` folder of the project. Not sure whether this is the correct location.

This is my appsettings.json:

```
{
    "BootstrapBlazorOptions": {
      "DefaultCultureInfo": "de"
    }
}
```


This is the code of MauiProgram.cs:

```
public static class MauiProgram
{
	public static MauiApp CreateMauiApp()
	{
		var dbFactory = new AppContextFactory();
		dbFactory.EnsureMigratedDatabase(DatabaseFilePath);

                var builder = MauiApp.CreateBuilder();
		builder
			.UseMauiApp<App>()
			.ConfigureFonts(fonts =>
			{
				fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
			});

		builder.Services.AddMauiBlazorWebView();
		builder.Services.AddLocalization();

		builder.Services.AddBootstrapBlazor(
			op =>
			{
				op.DefaultCultureInfo = "de-DE";
			},
			options =>
			{
				options.IgnoreLocalizerMissing = false;
				options.AdditionalJsonAssemblies = new Assembly[]
				{
					typeof(MauiProgram).Assembly
				};
			}
		);

                builder.Services.AddDbContext<AppDbContext>(options =>
			options.UseSqlite($"Data Source={DatabaseFilePath}")
		);

#if DEBUG
        	builder.Services.AddBlazorWebViewDeveloperTools();
		builder.Logging.AddDebug();
#endif

                CultureInfo.DefaultThreadCurrentCulture = new CultureInfo("de-DE");
                CultureInfo.DefaultThreadCurrentUICulture = new CultureInfo("de-DE");

		return builder.Build();
	}
}
```

**Reply by densen2014** [reply]
> appsettings.json

1. where is your `appsettings.json` ? better upload a demo.zip, because appsettings.json in maui is very diffent with other .net core projet

**Reply by tdittmar** [reply]
It's actually in my project's main folder. I'll try to assemble a "dummy project" illustrating the issue when I'm home from work. OTOH I'm using above code to set things, too - shouldn't that work regardless of whether appsettings.json is in the correct place?

**Reply by densen2014** [reply]
> 它实际上在我项目的主文件夹中。当我下班回家时，我会尝试组装一个“虚拟项目”来说明这个问题。OTOH 我也使用上面的代码来设置东西 - 无论 appsettings.json 是否在正确的位置，它都不应该工作吗？

in maui it has two ways, appsettings.json load with MauiAsset,  like `using var stream = a.GetManifestResourceStream("BlazorMaui.appsettings.json");`,  or `var stream = LoadMauiAsset().Result;  if (stream!=null) builder.Configuration.AddJsonStream(stream);` 

maybe you can not load json correct in you projet.  for example:  `Locales\de.json`. so we need your mini-demo project to review it


```
        async static Task<Stream?> LoadMauiAsset()
        {
            try
            {

                using var stream = await FileSystem.OpenAppPackageFileAsync("appsettings.json");
                using var reader = new StreamReader(stream);

                var contents = reader.ReadToEnd();
                Console.WriteLine("OpenAppPackageFileAsync => " + contents);
                var memoryStream = new MemoryStream();
                stream.CopyTo(memoryStream);
                return memoryStream;
            }
            catch (Exception e)
            {
                Console.WriteLine("OpenAppPackageFileAsync Exception => " + e.Message);
            }
            return null;
        }
```

----

```
            try
            {
                var a = Assembly.GetExecutingAssembly();
                using var stream = a.GetManifestResourceStream("BlazorMaui.appsettings.json");
                if (stream != null) builder.Configuration.AddJsonStream(stream);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
```

**Reply by tdittmar** [reply]
Yes, I'll create one that illustrates what I'm trying and come back to you. Thanks for looking into this so far!

**Reply by ArgoZhang** [reply, marked as answer]
### copy json localization file

copy https://github.com/dotnetcore/BootstrapBlazor/blob/main/localization/de.json file into your project `/Locales/de.json` and set build action to `Embedded resource`

<img width="390" alt="image" src="https://user-images.githubusercontent.com/22001478/235898804-76d7693d-ae61-459f-8915-4cedce0967ba.png">

### configuration

```
builder.Services.AddBootstrapBlazor(localizationConfigure: op =>
{
    op.AdditionalJsonAssemblies = new Assembly[]
    {
        typeof(Main).Assembly,
    };
});

builder.Services.Configure<BootstrapBlazorOptions>(op =>
{
    op.DefaultCultureInfo = "de";
    op.SupportedCultures = new List<string> { "de" };
});

var culture = new CultureInfo("de-DE");
CultureInfo.DefaultThreadCurrentCulture = culture;
CultureInfo.DefaultThreadCurrentUICulture = culture;    
```

before `return builder.Build();`

### Main.razor

```
<BootstrapBlazorRoot>
    <Router AppAssembly="@typeof(App).Assembly">
        <Found Context="routeData">
            <PageTitle>Title</PageTitle>
            <RouteView RouteData="@routeData" DefaultLayout="@typeof(MainLayout)" />
            <FocusOnNavigate RouteData="@routeData" Selector="h1" />
        </Found>
        <NotFound>
            <PageTitle>Not found</PageTitle>
            <LayoutView Layout="@typeof(MainLayout)">
                <p>正在玩命开发中 ...</p>
            </LayoutView>
        </NotFound>
    </Router>
</BootstrapBlazorRoot>
```

wrapper with component `BootstrapBlazorRoot`

### Test de localization

```
@page "/"
@using Microsoft.Extensions.Localization;
@inject IStringLocalizer<BootstrapBlazor.Components.AutoComplete> Localizer

<h1>Hello, world!</h1>

Welcome to your new app.

<SurveyPrompt Title="How is Blazor working for you?" />

<div>@Test</div>

@code {
    private string? Test { get; set; }

    protected override void OnInitialized()
    {
        base.OnInitialized();

        Test = Localizer["PlaceHolder"];
    }
}
```

in the Index page you will see `Bitte eingeben`

this is demo link https://github.com/ArgoZhang/MauiApp1

**Reply by tdittmar** [reply]
I'm excited now and will try this right away when I get home!

**Reply by tdittmar** [reply]
It works!! Thank you so much for your help. With some adjustments to the date formatting rules (should be dd\\.MM\\.yyyy in German locale) I get what I expected!

**Reply by densen2014** [reply]
``` 
var cultureInfo = new CultureInfo("es");
cultureInfo.NumberFormat.CurrencySymbol = "€";
cultureInfo.NumberFormat.NumberDecimalSeparator = ",";
cultureInfo.NumberFormat.CurrencyDecimalSeparator = ",";

CultureInfo.DefaultThreadCurrentCulture = cultureInfo;
CultureInfo.DefaultThreadCurrentUICulture = cultureInfo;

```
